x-pack/plugin/apm/src/internalClusterTest/java/org/elasticsearch/xpack/apm/APMRolloverIT.java

```diff
@@ -0,0 +1,166 @@
+
+    private ClusterService clusterService;
+    private APMIndexTemplateRegistry apmIndexTemplateRegistry;
+    private StackTemplateRegistry stackTemplateRegistry;
+    private ILMHistoryTemplateRegistry ilmHistoryTemplateRegistry;
+
+    @Before
+    public void obtainRegistry() {
+        clusterService = internalCluster().getInstance(ClusterService.class, internalCluster().getMasterName());
+        apmIndexTemplateRegistry = internalCluster().getInstance(APMIndexTemplateRegistry.class, internalCluster().getMasterName());
+        stackTemplateRegistry = internalCluster().getInstance(StackTemplateRegistry.class, internalCluster().getMasterName());
+        ilmHistoryTemplateRegistry = internalCluster().getInstance(ILMHistoryTemplateRegistry.class, internalCluster().getMasterName());
+    }
+
+    @After
+    public void unregisterRegistries() {
+        // we must remove the registries from the cluster state listeners' list, otherwise they mess up the cluster cleanup assertions by
+        // automatically reinstalling resources
+        clusterService.removeListener(apmIndexTemplateRegistry);
+        clusterService.removeListener(stackTemplateRegistry);
+        clusterService.removeListener(ilmHistoryTemplateRegistry);
+        // todo: find a way to cancel the geoip database download task as it throws errors when terminating the test
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return List.of(
+            APMPlugin.class,
+            StackPlugin.class,
+            IndexLifecycle.class,
+            Wildcard.class,
+            DataStreamsPlugin.class,
+            XPackPlugin.class,
+            IngestCommonPlugin.class,
+            IngestGeoIpPlugin.class,
+            IngestUserAgentPlugin.class,
+            PainlessPlugin.class,
+            MapperExtrasPlugin.class,
+            ConstantKeywordMapperPlugin.class,
+            AggregateMetricMapperPlugin.class,
+            AnalyticsPlugin.class
+        );
+    }
+
+    public void testRollover() throws Exception {
+        Set<String> apmIndexTemplates = new HashSet<>(apmIndexTemplateRegistry.getComposableTemplateConfigs().keySet());
+        ClusterState state = clusterService.state();
+        apmIndexTemplateRegistry.clusterChanged(new ClusterChangedEvent(APMRolloverIT.class.getName(), state, state));
+        // No need to waste the initial frequent assertBusy iterations because this takes a few seconds at least
+        Thread.sleep(3000);
+        assertBusy(() -> {
+            // todo - figure out why we need this synthetic cluster change event trigger to make sure the last index template is registered
+            // maybe because org.elasticsearch.ingest.IngestService.putPipeline() doesn't update the cluster state and trigger state
+            // change event like, for example, org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.addComponentTemplate does
```

Comment:
My hypothesis is that in this case we don't install the [`metrics-apm.internal` index template](https://github.com/elastic/elasticsearch/blob/bac147f54cd5845d8d1593d9657f39917de610c9/x-pack/plugin/apm/src/main/resources/index-templates/metrics-apm.internal.yaml), because it depends on the [`metrics-apm.internal` component template](https://github.com/elastic/elasticsearch/blob/bac147f54cd5845d8d1593d9657f39917de610c9/x-pack/plugin/apm/src/main/resources/component-templates/metrics-apm.internal.yaml), which in turn depends on the [`metrics-apm.internal` ingest pipeline](https://github.com/elastic/elasticsearch/blob/bac147f54cd5845d8d1593d9657f39917de610c9/x-pack/plugin/apm/src/main/resources/ingest-pipelines/metrics-apm.internal.yaml) when the cluster-change-event gets processed.

@jbaiera do you think we missed this in https://github.com/elastic/elasticsearch/pull/95782? I don't think installation of ingest pipeline triggers a cluster change event, and for the registries to keep installing everything, it should.

Reply:
Ingest pipeline installation always triggers a cluster state update if the cluster is able to accept an ingest workload:

https://github.com/elastic/elasticsearch/blob/a20ee3f8f268a4f181b3e898635a96a3f605721a/server/src/main/java/org/elasticsearch/ingest/IngestService.java#L431-L435

Reply:
Looking at the log of this test, the component template's is being added properly, but it may not be triggering the state change. See what happens if I remove the manual state change event (notice the timings):
```
[2023-09-13T04:18:15,091][INFO ][o.e.x.c.t.IndexTemplateRegistry] [node_t0] adding ingest pipeline metrics-apm.internal
[2023-09-13T04:18:15,093][INFO ][o.e.c.m.MetadataIndexTemplateService] [node_t0] adding component template [metrics-apm.internal]
[2023-09-13T04:18:27,224][INFO ][o.e.x.a.APMRolloverIT    ] [testRollover] [APMRolloverIT#testRollover]: cleaning up after test
```

And this is how it looks when I trigger such event within the `assertBusy` block, like the current code does:
```
[2023-09-13T01:22:55,447][INFO ][o.e.x.c.t.IndexTemplateRegistry] [node_t0] adding ingest pipeline metrics-apm.internal
[2023-09-13T01:22:55,449][INFO ][o.e.c.m.MetadataIndexTemplateService] [node_t0] adding component template [metrics-apm.internal]
[2023-09-13T01:22:57,714][INFO ][o.e.c.m.MetadataIndexTemplateService] [node_t0] adding index template [metrics-apm.internal] for index patterns [metrics-apm.internal-*]
```

There is definitely something wrong here, I'll debug later and see. If you have an idea- let me know

Reply:
The issue went away after @axw refactored the components, so I removed the manual event trigger and resolve this thread, but I mean to restore the former state and figure out what the problem is. I think there is a real issue with cluster state events being missed in some configurations.